resolwe_bio/processes/expression/star_quantification.py

```diff
@@ -0,0 +1,509 @@
+    with open(file=input_file, mode="rb") as f_in:
+        with gzip.open(filename=output_file, mode="wb") as f_out:
+            shutil.copyfileobj(f_in, f_out)
+
+
+def exp_to_df(exp_file, exp_type):
+    """Prepare expression file for gene sets merging."""
+    with open(exp_file) as exp:
+        df = pd.read_csv(exp, sep="\t", float_precision="round_trip")
+        df.rename(
+            index=str,
+            columns={
+                "EXPRESSION": exp_type,
+            },
+            inplace=True,
+        )
+        # Cast FEATURE_ID column to string
+        df["FEATURE_ID"] = df["FEATURE_ID"].astype("str")
+
+    return df
+
+
+def prepare_expression_set(rc, tpm, cpm, feature_dict, outfile_name):
+    """Prepare expression set output data."""
+    rc_exp = exp_to_df(rc, "RAW_COUNT")
+    tpm_exp = exp_to_df(tpm, "TPM")
+    cpm_exp = exp_to_df(cpm, "CPM")
+    rc_exp["GENE_SYMBOL"] = rc_exp["FEATURE_ID"].map(feature_dict)
+    input_features = rc_exp["FEATURE_ID"].tolist()
+    # Check if all of the input feature IDs could be mapped to the gene symbols
+    if not all(f_id in feature_dict for f_id in input_features):
+        print(
+            f"{sum(rc_exp.isnull().values.ravel())} feature(s) "
+            f"could not be mapped to the associated feature symbols."
+        )
+    # Merge with normalized expression values
+    exp_set = rc_exp.merge(tpm_exp, on="FEATURE_ID")
+    exp_set = exp_set.merge(cpm_exp, on="FEATURE_ID")
+    # Reorder columns
+    columns = ["FEATURE_ID", "GENE_SYMBOL", "RAW_COUNT", "TPM", "CPM"]
+    exp_set = exp_set[columns]
+    # Replace NaN values with empty string
+    exp_set.fillna("", inplace=True)
+
+    # Write to file
+    exp_set.to_csv(
+        outfile_name + ".txt.gz",
+        header=True,
+        index=False,
+        sep="\t",
+        compression="gzip",
+    )
+
+    # Write to JSON
+    df_dict = exp_set.set_index("FEATURE_ID").to_dict(orient="index")
+    with open(outfile_name + ".json", "w") as f:
+        json.dump({"genes": df_dict}, f, allow_nan=False)
+
+
+def expression_to_storage(rc_input, rc_output):
+    """Convert expressions file to JSON format."""
+
+    def isfloat(value):
+        """Check if value is float."""
+        try:
+            float(value)
+            return True
+        except ValueError:
+            return False
+
+    with io.TextIOWrapper(io.BufferedReader(gzip.open(rc_input))) as f:
+        # Split lines by tabs
+        # Ignore lines without a number in second column
+        # Build a dictionary of gene-expression pairs
+        exp = {
+            "genes": {
+                gene_exp[0]: float(gene_exp[1])
+                for gene_exp in (l.split("\t") for l in f)
+                if len(gene_exp) == 2 and isfloat(gene_exp[1])
+            }
+        }
+
+    with open(file=rc_output, mode="wt") as f:
+        json.dump(exp, f)
+
+    return rc_output
+
+
+class STARQuantification(ProcessBio):
+    """STAR quantification.
```

Comment:
I suggest the following description: `Normalize STAR quantification results`. "STAR quantification" wrongly implies that the full quantification by STAR aligner (while aligning) is done here. Modify the class name accordingly.